---
title: "Cell type deconvolution with BisqueRNA"
author: "RA ([numpde](https://github.com/numpde/))"
date: "2020-Dec-24"
output: html_document
editor_options:
 chunk_output_type: console
---


## Pertinent links

- [Source code](https://github.com/numpde/sta426_deconvo/tree/main/code/sandbox_ra/20201222-Bisque),
  short link: 
  [http://bit.ly/b_bisque](http://bit.ly/b_bisque)

- [FGCZ bulk dataset](https://github.com/numpde/sta426_deconvo/tree/main/data/20201128-FGCZ)

- [Darmanis et al (2015)](https://github.com/numpde/sta426_deconvo/tree/main/data/20201206-RefDatasets/2015-Darmanis) 
  reference scRNA dataset

- [Allen Brain M1](https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x) 
  reference scRNA dataset

- [Analysis](https://github.com/numpde/sta426_deconvo/tree/main/code/sandbox_ra/20201220-Darmanis_vs_Allen) 
  of **Darmanis et al** in terms of **Allen Brain M1**

- Bisque 
  [paper](https://www.nature.com/articles/s41467-020-15816-6),
  [vignette](https://cran.r-project.org/web/packages/BisqueRNA/vignettes/bisque.html),
  [manual](https://cran.r-project.org/web/packages/BisqueRNA/BisqueRNA.pdf),
  [github](https://github.com/cozygene/bisque),
  [ReferenceBasedDecomposition](https://rdrr.io/cran/BisqueRNA/src/R/reference_based.R)

- Bisque uses 
  [limSolve](https://cran.r-project.org/web/packages/limSolve/limSolve.pdf)



## Preliminaries {.tabset .tabset-fade .tabset-pills}

### Setup    

```{r setup, include=TRUE}
suppressPackageStartupMessages({
  library(magrittr)
  library(dplyr)
  library(ggplot2)
  library(pheatmap)
  library(stringr)
  library(reshape2)
  library(rlang)

  # https://vroom.r-lib.org/articles/vroom.html
  # isntall.packages("vroom")
  library(vroom)

  # install.packages("pathlibr")
  library(pathlibr)

  # Provides `ExpressionSet` data structure
  # BiocManager::install("Biobase")
  library(Biobase)

  # Deconvolution package
  # install.packages("BisqueRNA")
  library(BisqueRNA)
})
```

### Paths

```{r}
BASEPATH <- pathlibr::Path$new(".")
```

```{r}
out_file <- (function(.) BASEPATH$join("./b_output")$join(.)$show)
dir.create(out_file(""), showWarnings = FALSE)
```

```{r}
path_to <- (function(.) Sys.glob(BASEPATH$join("../../..")$join(.)$show))
```

### Random-seed

```{r}
set.seed(43)
```

### Row names, etc.

```{r}
# Use first column as index
by_col1 <- (function(.) tibble::column_to_rownames(., colnames(.)[1]))
# Use index as new column `name`
ind2col <- (function(., name) tibble::rownames_to_column(., var = name))
```

```{r}
group_by_and_sum <-
  function(., column_name) {
    (.) %>%
      dplyr::group_by(!!rlang::sym(column_name)) %>%
      dplyr::summarise_all(sum) %>%
      tibble::column_to_rownames(column_name)
  }
```

```{r}
fix_names <-
  function(.) {
    (.) %>%
      # https://stackoverflow.com/a/55184433
      set_names(names(.) %>% str_replace("cell type", "celltype"))
  }
```

### Tab-delimited files

Note: The function `utils::read.delim` is too slow
to read wide tables.

```{r}
Sys.setenv("VROOM_CONNECTION_SIZE" = 131072 * 100)
from_csv <- (function(.) suppressMessages(vroom(., del = "\t")) %>% by_col1())
```

A call to `to_csv` should be preceded by
`ind2col("name for index")`.

```{r}
to_csv <- (function(., f) vroom_write(., out_file(f), delim = "\t"))
```

### Plotting

```{r}
ggplot2::theme_set(theme_light(base_size = 15))
```

```{r}
# Wide plots (inches)
WIDTH1 <- 20
HEIGHT1 <- 3
```

```{r}
hist_theme <-
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.ticks.x = element_blank(),
  )
```

```{r}
save_barchart <-
  function(., filename) {
    as.data.frame(.) %>%
      ind2col("celltype") %>%
      reshape2::melt(
        id = "celltype",
        var = "sample",
        value.name = "y"
      ) %>%
      {
        ggplot(., aes(x = sample, y = y, fill = celltype)) +
          geom_bar(stat = "identity") +
          scale_fill_brewer(palette = "Paired") +
          ylim(0, 1) +
          hist_theme +
          theme(
            legend.title = element_blank(),

            axis.text.y = element_blank(),

            axis.title.x = element_blank(),
            axis.text.x = element_text(
              a = -90,
              vjust = 0.5,
              hjust = 0,
              size = 10,
            )
          ) +
          ggsave(
            filename = out_file(filename),
            width = WIDTH1,
            height = HEIGHT1,
            device = "png"
          )
      }
    return(.)
  }

# Example:
# fgcz_darm %>% save_barchart("fgcz_darm_data_bars.png")
```

```{r}
save_heatmap <-
  function(., filename) {
    as.data.frame(.) %>%
      pheatmap::pheatmap(
        filename = out_file(filename),
        cluster_rows = FALSE,
        cluster_cols = FALSE,
        fontsize_row = 14,
        fontsize_col = 10,
        width = WIDTH1,
        height = HEIGHT1
      )
    return(.)
  }
```


## Data sources {.tabset .tabset-fade .tabset-pills}


### Bulk expression: FGCZ

```{r}
fgcz_data <-
  path_to("data/20201128-FGCZ/*count.zip") %>%
  unz(., unzip(., list = TRUE)$Name) %>%
  from_csv() %>%
  # Collapse ENSG IDs by gene_name:
  group_by_and_sum("gene_name")

fgcz_meta <- from_csv(path_to("data/20201128-FGCZ/*infos.tsv"))
```


### Ref scRNA: Darmanis

```{r}
darm_data <- from_csv(path_to("data/*/2015-Darmanis/b*/data.csv.gz"))
darm_meta <- from_csv(path_to("data/*/2015-Darmanis/b*/meta.csv.gz")) %>%
  fix_names()
```

```{r}
exclude_celltypes <- c("fetal_quiescent", "fetal_replicating", "hybrid")
```

Plot cell type counts.

```{r}
darm_meta$celltype %>%
  data.frame(x = .) %>%
  ggplot(aes(x = x, fill = if_else(x %in% exclude_celltypes, "drop", "keep"))) +
  geom_bar() +
  ggtitle("Cell types in the 'Darmanis' reference dataset") +
  scale_y_log10() +
  hist_theme +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x = element_text(a = 45, hjust = 1)) +
  theme(legend.title = element_blank())
```

Remove unnecessary cell types
from the scRNA reference dataset.

```{r}
darm_meta <-
  darm_meta %>%
  dplyr::filter(!(celltype %in% exclude_celltypes))

darm_data <-
  darm_data %>%
  dplyr::select_if(names(.) %in% rownames(darm_meta))

stopifnot(285 == nrow(darm_meta))
stopifnot(285 == ncol(darm_data))
```


### Ref scRNA: Allen Brain M1

```{r}
abm1_meta <- from_csv(path_to("data/*/2019-AllenBrain-M1/b*/meta.csv*"))
abm1_data <- from_csv(path_to("data/*/2019-AllenBrain-M1/b*/data.csv*"))
```

Some sanity checks.

```{r}
# Number of single cells
stopifnot(ncol(abm1_data) > 70000)
stopifnot(nrow(abm1_meta) > 70000)
# Number of genes and some examples
stopifnot(nrow(abm1_data) == 141)
stopifnot(all(c("PIK3CD", "WNT4", "LDLRAP1") %in% rownames(abm1_data)))
```

Drop samples with too little expression,
otherwise Bisque doesn't go through.

```{r}
abm1_data <-
  abm1_data %>%
  # Keep only genes common with the FGCZ dataset
  {
    (.)[rownames(.) %in% rownames(fgcz_data), ]
  } %>%
  # Drop zero columns (dplyr variant too slow)
  {
    (.)[, colSums(.) != 0]
  }

abm1_meta <-
  abm1_meta %>%
  dplyr::filter(rownames(.) %in% colnames(abm1_data))
```

Plot cell type counts.

```{r}
data.frame(
  x = abm1_meta$celltype,
  donor = abm1_meta$donor
) %>%
  ggplot(aes(x = x, fill = donor)) +
  geom_bar(position = "dodge") +
  ggtitle("Cell types in the 'Allen Brain M1' reference dataset") +
  scale_fill_brewer(palette = "Set1") +
  scale_y_log10() +
  hist_theme +
  theme(axis.title.x = element_blank()) +
  theme(axis.title.y = element_blank())
```


### Marker genes

These marker genes 
[were extracted](https://github.com/numpde/sta426_deconvo/tree/main/data/20201206-RefDatasets/2019-AllenBrain-M1)
from the
[Allen Brain M1](https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x)
scRNA dataset.
Technically, these should be the same as 
in `abm1_data` before filtering.
Just in case, we subset them the maximal common subset.


```{r}
abm1_markergenes <-
  path_to("data/*/*AllenBrain-M1/b*/marker_genes.csv") %>%
  from_csv() %>%
  rownames() %>%
  {
    (.)[(.) %in% rownames(fgcz_data)]
  } %>%
  {
    (.)[(.) %in% rownames(abm1_data)]
  }

stopifnot(117 == length(abm1_markergenes))
```


## Deconvolution by Bisque


### Deconvolution (from Darmanis) {.tabset .tabset-fade .tabset-pills}

In this section we use
**Darmanis et al**
as the reference dataset.

#### Bisque {#fgcz_darm}

Repackage the data
following the 
[Bisque vignette](https://cran.r-project.org/web/packages/BisqueRNA/vignettes/bisque.html).

```{r}
fgcz_eset <- Biobase::ExpressionSet(
  # Subset to marker genes for residual norm computation
  assayData = as.matrix(fgcz_data[abm1_markergenes, ])
)
```

```{r}
darm_eset <- Biobase::ExpressionSet(
  # Expression data
  assayData = as.matrix(darm_data),
  # Metadata
  phenoData = Biobase::AnnotatedDataFrame(
    data = data.frame(
      row.names = rownames(darm_meta),
      cellType = darm_meta$celltype,
      SubjectName = darm_meta$experiment_sample_name,
      check.names = FALSE,
      check.rows = FALSE,
      stringsAsFactors = FALSE
    ),
    varMetadata = data.frame(
      row.names = c("cellType", "SubjectName"),
      labelDescription = c("cellType", "SubjectName")
    )
  )
)
```

Deconvolution.

```{r}
bisque_report <-
  BisqueRNA::ReferenceBasedDecomposition(
    # BULK DATA
    bulk.eset = fgcz_eset,
    # REFERENCE
    sc.eset = darm_eset,
    #
    use.overlap = FALSE,
    markers = abm1_markergenes,
    verbose = FALSE
  )

# Fields of the Bisque result
print(as.data.frame(names(bisque_report)))
```

Bisque returns proportions 
that sum to one:

```{r}
stopifnot(max(abs(1 - colSums(bisque_report$bulk.props))) <= 1e-10)
```

We undo that using the residual.
The meaning of `rnorm` is not clear
from the
[manual](https://cran.r-project.org/web/packages/BisqueRNA/BisqueRNA.pdf),
so this might be incorrect:

```{r}
explained_fractions <-
  1 - (sqrt(bisque_report$rnorm) / sqrt(colSums(fgcz_data[bisque_report$genes.used, ]**2)))
```

Rescale proportions to 
the "explained fractions".

```{r}
fgcz_darm <-
  t(t(bisque_report$bulk.props) * explained_fractions)
```

Cluster bulk samples by composition.

```{r}
fgcz_darm <-
  fgcz_darm %>% {
    (.)[, hclust(dist(t(.)))$order]
  }
```

Save to disk
and
visualize inferred cell type composition
by bulk sample.

```{r, results = FALSE}
fgcz_darm %>%
  as.data.frame() %>%
  save_heatmap("fgcz_darm_heat.png") %>%
  save_barchart("fgcz_darm_bars.png") %>%
  ind2col("celltype") %>%
  to_csv("fgcz_darm_data.csv")
```

![](`r out_file("fgcz_darm_heat.png")`)
![](`r out_file("fgcz_darm_bars.png")`)


#### Baseline NNLS

For the code see 
[here](https://github.com/numpde/sta426_deconvo/blob/main/code/sandbox_ra/20201206-NNLS_Darmanis/a_deconvolution0.py).

```{r}
fgcz_nnls <-
  path_to("code/*/*-NNLS_Darmanis/a_*/celltypes.csv") %>%
  # The first column is not unique
  # The header contains minuses, e.g. "H-0C083", keep them
  utils::read.delim(., check.names = F) %>%
  # `cell type` ~> `celltype`
  fix_names() %>%
  # Collapse cell types:
  group_by_and_sum("celltype") %>%
  # Order columns in the same way:
  dplyr::relocate(colnames(fgcz_darm))
```

Visualize and save to disk.

```{r}
fgcz_nnls %>%
  as.data.frame() %>%
  save_heatmap("fgcz_nnls_heat.png") %>%
  save_barchart("fgcz_nnls_bars.png") %>%
  ind2col("celltype") %>%
  to_csv("fgcz_nnls_data.csv")
```

![](`r out_file("fgcz_nnls_heat.png")`)
![](`r out_file("fgcz_nnls_bars.png")`)



### Deconvolution (from Allen Brain M1)

In this section we use
**Allen Brain M1**
as the reference dataset.

Repackage:

```{r}
abm1_eset <- Biobase::ExpressionSet(
  # Expression data
  assayData = as.matrix(abm1_data),
  # Metadata
  phenoData = Biobase::AnnotatedDataFrame(
    data = data.frame(
      row.names = rownames(abm1_meta),
      cellType = abm1_meta$celltype,
      SubjectName = abm1_meta$donor,
      check.names = FALSE,
      check.rows = FALSE
    ),
    varMetadata = data.frame(
      row.names = c("cellType", "SubjectName"),
      labelDescription = c("cellType", "SubjectName")
    )
  )
)
```

Deconvolution:

```{r}
bisque_report <-
  BisqueRNA::ReferenceBasedDecomposition(
    # BULK DATA
    bulk.eset = fgcz_eset,
    # REFERENCE
    sc.eset = abm1_eset,
    #
    use.overlap = FALSE,
    markers = abm1_markergenes,
    verbose = FALSE
  )
```

Order columns consistently with above attempt.
  
```{r}
fgcz_abm1 <-
  bisque_report$bulk.props %>%
  as.data.frame() %>%
  dplyr::relocate(colnames(fgcz_darm))
```

Visualize and save to disk.

```{r}
fgcz_abm1 %>%
  as.data.frame() %>%
  save_heatmap("fgcz_abm1_heat.png") %>%
  save_barchart("fgcz_abm1_bars.png") %>%
  ind2col("celltype") %>%
  to_csv("fgcz_abm1_data.csv")
```

![](`r out_file("fgcz_abm1_heat.png")`)
![](`r out_file("fgcz_abm1_bars.png")`)


***
```{r include=FALSE}
knitr::knit_exit()
```


The following is under construction.


### Discussion

The cell type proportions [returned by Bisque](#fgcz_darm)
vary more than expected.
Let us introduce 
the deviations from the average predicted profile
as

```{r}
fgcz_darm_mean <- 
```

and the deviations from it as

```{r}
fgcz_darm_dev <- (fgcz_darm - rowMeans(fgcz_darm))
```


```{r}
data.frame(
  q = explained_fractions,
  r = fgcz_meta$RIN,
  s = fgcz_meta$Source
) %>%
  ggplot(aes(x = q, y = r, color = s)) +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Explained fraction", y = "RIN", color = "Sample source") +
  geom_point(size = 5)
```



[//]: # XX -- TODO

```{r, include=FALSE, echo=FALSE}
library(SingleCellExperiment)
library(scater)
factor_name <- "Condition"
fgcz_data %>%
  edgeR::cpm(.) %>%
  {
    (.)[abm1_markergenes, ]
  } %>%
  {
    list(logcounts = log(1 + .))
  } %>%
  SingleCellExperiment(assays = .) %>%
  scater::runMDS() %>%
  SingleCellExperiment::reducedDim() %>%
  as.data.frame() %>%
  setNames(c("x", "y")) %>%
  ggplot(aes(x = x, y = y, color = fgcz_meta[[factor_name]])) +
  labs(color = factor_name) +
  geom_point(size = 5)
```

```{r, include=FALSE, echo=FALSE}
fgcz_abm1 %>%
  t() %>%
  as.data.frame() %>%
  dplyr::mutate(condition = fgcz_meta$Condition) %>%
  {
    MASS::lda(formula = condition ~ ., data = (.))
  }
```
