---
title: "Cell-type deconvolution"
author: "Kalvin Dobler"
output: html_document
---

**Setting up the environment and importing libraries**\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
rm(list = ls())
library(dplyr)
library(MASS)
library(tidyverse)
library(edgeR)
library(limma)
library(devtools)
library(markerGeneProfile)
library(dtangle)
library(glmnet)
library(MineICA)
library(amritr)
library(mixOmics)
library(preprocessCore)
library(DSA)
library(UpSetR)
library(GEOquery)
library(reshape2)
library(pheatmap)
library(RColorBrewer)
library(ADAPTS)
```

***
***

##Exploratory data analysis of the data

**Loading of the data**\
The dataset consists of the expression values of more than 20'000 genes from 102 samples, including 52 patients with amyotrophic lateral sclerosis (ALS) and 50 controls. This disease is known to involve the nerve cells responsible for controlling voluntary muscle movement [1]. Hence, the aim of the following analysis is to deconvolve the tissue heterogeneity from bulk RNA-seq in order to get indication of changes in cell-type proportions between the two distinct conditions. Cell-type deconvolution has emerged over the past decade [2] as a recognized computational method to analyze cell subtype proportion between samples. Indeed, this variability between samples can be caused by underlying variation of genomic properties such as gene expression or DNA methylation [3]. It can thus be of great interest to analyze the phenotypic variations in order to gain insight of potential genomic alterations.\
\
To start, the dataset of expression values is converted into a matrix in which the row names correspond to genes. Since dupplicated reads are present, only the reads with higher expression are selected.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
data <- read.table('Count_QC-raw-count.txt', sep = '\t', header = TRUE)

data_unique <- data[-c(8382, 5347, 21408, 21409, 21406, 3914, 21412, 21412, 21402, 21403, 21410, 20201, 21399,10000,21404,
                       21463, 5918, 21407, 3036, 8486, 21398, 7989, 21400, 2706, 21401, 21405, 6682, 7324, 21461, 1982,
                       21051, 21462, 21411, 3706),]

rownames(data_unique) <- data_unique[,2]

data_unique <- data_unique[,3:ncol(data_unique)]
```

***

**Data exploration**\
The following procedure follows the package edgeR [4]. The data are then filtered and normalized. In particular, it is worth removing genes with constantly low counts as they are unlikely to be differentially expressed. The multidimensional scaling plot below depicts the distance between samples. At first sight, the two dimensions do not seem to separate the two conditions.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
group <- c(rep(1,52),rep(2,50))
group2 <- c(rep('ALS',52),rep('Control',50))

y <- DGEList(counts = data_unique, group = group)

keep <- filterByExpr(y)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y2 <- cpm(y, log = T)

plotMDS(y, col = c('tomato', 'cornflowerblue')[as.factor(group)])
legend(1.1, -0.9, legend=c("ALS", "Control"),col=c("tomato", "cornflowerblue"),
       lty = 1:1, bty = "n", cex=1)
```

***

**Visualization of the expression values**\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
y_df <- as.data.frame(y)
rownames(y_df) <- rownames(y)
df_reshaped <- melt(round(log1p(y_df[,1:102]),2), value.name = "value")

ggplot(df_reshaped, aes(x = variable, y = value)) +
  geom_violin(aes(fill = variable) ,trim = T) +
  ggtitle("Violin plots of the expression values") +
  theme_light(base_size = 20) +
  theme(axis.text.x = element_text(angle = 90, size = 4),
        legend.position = 'none', plot.title = element_text(size =20))
```

***

**Estimation of the dispersion**\
The dispersion for the dataset is estimated and it returns the coefficient of biological variation which is 0.3.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
design <- model.matrix(~factor(group))

y <- estimateDisp(y, design)

sqrt(y$common.dispersion)

plotBCV(y)
```

***

**Analysis of the differential expression**\
Then, differentially expressed genes are computed. The plot below depicts the log-fold change against log-counts per million. The table summarizes the results.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
fit <- glmFit(y,design)

lrt <- glmLRT(fit)

#topTags(lrt)
#top_order <- order(lrt$table$PValue)
#cpm(y)[top_order[1:10],]

plotMD(lrt)

knitr::kable(summary(decideTests(lrt)), format = "simple", align = "c", col.names = "Summary")
```

***

**Principal component analysis**\
A principal component analysis is computed. The two major axes explain approximately 40% of the total variance. Again, there is no clear distinction between groups. As one might notice, the control group seems to be more scattered than the ALS one.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
pca <- prcomp(t(y2), scale = F)
var_explained <- pca$sdev^2/sum(pca$sdev^2)

pca$x %>% 
  as.data.frame %>%
  ggplot(aes(x=PC1,y=PC2, group = group)) + 
  geom_point(aes(x=PC1,y=PC2,colour=factor(group2)),size=2) +
  theme_light(base_size = 20) + 
  labs(x=paste0("PC1: ",round(var_explained[1]*100,1),"%"),
       y=paste0("PC2: ",round(var_explained[2]*100,1),"%"),
       color = "Groups") +
  scale_fill_discrete(name = "Groups", labels = c("ALS", "Control")) +
  ggtitle('Principal Component Analysis')
```

***

**Visualization of the top 500 genes with clustered heatmap**\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
top_genes <- rownames(topTags(lrt, n = 500))

y_df2 <- y_df[row_number(top_genes),]
rownames(y_df2) <- top_genes

pheatmap(y_df2, scale = 'row', cluster_rows = F,show_rownames = F,
         fontsize_col = 4)
```

***

**Distance between samples**\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
sampleDists <- dist(t(y_df2), method = "euclidean")
sampleDistMatrix <- as.matrix(sampleDists)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,
         fontsize_row = 4,
         fontsize_col = 4)
```

***
***

##Prepocessing and analysis of the data

**Filtering and normalization of the dataset**\
The dataset is prepared for cell-type deconvolution.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
group <- c(rep(1,52),rep(2,50))

dge <- DGEList(counts = data_unique, group = group)

keep <- filterByExpr(dge)

dge <- dge[keep, , keep.lib.sizes=FALSE]

geneExpr <- cpm(dge)

geneExpr <- as.data.frame(geneExpr)
```

***

**Loading the dataset of marker genes**\
To adapt the following deconvolution methods to brain tissue, the human brain cell-sorted RNA-seq data ("Darmanis") [5] is used to determine cell-type specific expression profiles of the marker genes. In the following, only neurons, oligodendrocytes, astrocytes, microglia, endothelial and oligodendrocyte precursor cells (OPC) are considered.\
\
The dataset is prepared in a way such that each row corresponds to a gene and each column to a specific cell type.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
data <- getGEO('GSE67835')

data1 <- data$`GSE67835-GPL15520_series_matrix.txt.gz`
data2 <- data$`GSE67835-GPL18573_series_matrix.txt.gz`

data1 <- pData(data1)
data2 <- pData(data2)

raw <- read.csv('data.csv', header = TRUE, sep = '\t')
rownames(raw) <- raw[,1]
raw <- raw[,-1]
raw <- as.data.frame(raw)

raw2 <- raw
colnames(raw2)[1:121] <- data2[1:121,45]
colnames(raw2)[122:135] <- data1[1:14,45]
colnames(raw2)[136:313] <- data2[122:299,45]
colnames(raw2)[314:437] <- data1[15:138,45]
colnames(raw2)[438:466] <- data2[300:328,45]

raw_filt <- raw2[,which(colnames(raw2) %in% c('oligodendrocytes','astrocytes','OPC','microglia','neurons','endothelial'))]
colnames(raw_filt) <- colnames(raw2)[which(colnames(raw2) %in% c('oligodendrocytes','astrocytes','OPC','microglia','neurons','endothelial'))]

raw3 <- raw[,which(colnames(raw2) %in% c('oligodendrocytes','astrocytes','OPC','microglia','neurons','endothelial'))]
```

***

**Searching for top differentially expressed genes**\
Next, differential gene expression analysis is performed in order to determine a set of specific marker genes for each cell-type.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
commongenes <- intersect(rownames(geneExpr), rownames(raw_filt))

dge2 <- DGEList(counts = raw_filt[commongenes,])

dge2 <- calcNormFactors(dge2)

ref_markers <- cpm(dge2)

markers_mean = sapply(unique(colnames(raw_filt)),function(x) rowMeans(raw_filt[,which(colnames(raw_filt)==x)]))

h <- as.data.frame(colnames(raw_filt))

rownames(h) <- colnames(raw3)

colnames(h) <- "cell_type"

f <- paste(colnames(h),colnames(raw_filt), sep = "_")

design <- model.matrix(~0+f)

rownames(design) <- rownames(h)

colnames(design) <- unique(f)

commongenes2 = intersect(rownames(geneExpr),rownames(markers_mean))

voom_matrix <- voom(raw[commongenes2,rownames(design)], design)

fit <- lmFit(voom_matrix, design)

contrast_name <- sapply(colnames(design),function(x)paste(x,'-(',paste(colnames(design)[colnames(design)!=x],collapse = "+"),')/5',sep = ''))

contrast <- makeContrasts(contrasts = contrast_name, levels = colnames(design))

fit2  <- contrasts.fit(fit, contrast)

fit2  <- eBayes(fit2)
```

***

**Visualization of intersecting sets**\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
d_test <- (decideTests(fit2) != 0)+0
d_test <- as.data.frame(d_test)
colnames(d_test) <- c("Oligodendrocytes", "Astrocytes", "OPC", "Microglia", "Neurons", "Endothelial")

upset(d_test)
```

***

**Selection of differentially expressed genes per cell type**\
For each cell-type, the top 100 differentially expressed genes are selected as marker genes.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
gene_markers <- NULL
for(i in 1:ncol(design)){
  x <- fit2$coef[p.adjust(fit2$p.v[,i],'fdr')<0.05,i]
  gene_markers[[colnames(design)[i]]] <-  names(head(sort(-x[x>0]),100))
}

knitr::kable(unlist(lapply(gene_markers,length)), format = "simple", align = "cccccc", col.names = "Number of marker genes")
```

***
***

##Cell-type deconvolution

The following deconvolution methods are classified as "supervised" as they rely on the presence of an external knowledge of cell-type specific expression profiles.\
\
**Deconvolution with dtangle**\
The method dtangle [6] requires two inputs of external knowledge: a reference data and marker genes that need to be combined in one matrix.
```{r, echo = TRUE, message = FALSE, warning = FALSE}
commongenes3 <- intersect(rownames(geneExpr), rownames(markers_mean))

geneExpr2 <- geneExpr[pmatch(commongenes3, rownames(geneExpr)),]

markers_mean2 <- markers_mean[pmatch(commongenes3, rownames(markers_mean)),]

Y <- cbind(markers_mean2, geneExpr2)

pure_samples <- as.list(1:6)
names(pure_samples) <- colnames(Y)[1:6]

dtDarmanis <- dtangle(log1p(t(Y)), pure_samples = pure_samples, markers = gene_markers)$estimates

# dt_csv <- dtDarmanis[-c(1:6),]
# dt_csv <- t(dt_csv)
# write.table(dt_csv, file = "Deconvolution_dtangle.csv", sep=" ")
```

***

**Visualization of the cell types' proportions**\
The plots below depicts the proportions (computed by the method) for each sample.\
```{r, echo = TRUE, message = FALSE, warning = FALSE, fig.show="hold", out.width="50%"}
dt <- dtDarmanis[-c(1:6),]

par(mar = c(6,4,4,8))

dt_reshaped1 <- melt(dt[1:52,], id = "columns")
dt_reshaped2 <- melt(dt[53:102,], id = "columns")

ggplot(dt_reshaped1, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
    scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
  theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('ALS')

ggplot(dt_reshaped2, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
  scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
    theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('Control')

```

***

**Deconvolution with non-negative least-squares**\
The method NNLS, initially proposed by Alexander R. Abbas et al. [7] and expands by Samuel A Danziger et al. [8], relies on finding the solution of a convolution equation of the form AX = B, where B is the dataset of expression values, X is the set of unknown cell-type proportions and A is the set of known marker genes per cell-type.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
prop <- estCellPercent.nnls(markers_mean2, geneExpr2)

prop <- prop[1:6,]

for (i in 1:ncol(prop)) {
  prop[,i] <- prop[,i]/colSums(prop)[i]
}

prop <- t(prop)

# prop_csv <- t(prop)
# write.table(prop_csv, file = "Deconvolution_NNLS.csv", sep=" ")
```

***

**Visualization of the cell types' proportions**\
```{r, echo = TRUE, message = FALSE, warning = FALSE, fig.show="hold", out.width="50%"}
par(mar = c(6,4,4,8))

prop_reshaped1 <- melt(prop[1:52,], id = "columns")
prop_reshaped2 <- melt(prop[53:102,], id = "columns")

ggplot(prop_reshaped1, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
  scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
  theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('ALS')

ggplot(prop_reshaped2, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
  scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
    theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('Control')

```

***

**Deconvolution with cibersort**\
The method cibersort [9] relies on the application of linear support vector regression (SVR).\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
Y <- DGEList(counts=Y)
Y <- calcNormFactors(Y,method = 'TMM')
Y <- cpm(Y, log=FALSE)

new_y <- Y[unlist(gene_markers),-c(1:6)]
new_y2 <- Y[unlist(gene_markers),1:6]
new_y2 <- (new_y2 - mean(new_y2))/sd(as.vector(new_y2))

csDarmanis <- sapply(as.data.frame(new_y),function(z) CoreAlg(new_y2,(z - mean(z))/sd(z))$w)
rownames(csDarmanis) <- colnames(new_y2)

csDarmanis <- t(csDarmanis)

# cs_csv <- t(csDarmanis)
# write.table(cs_csv, file = "Deconvolution_cibersort.csv", sep=" ")
```

***

**Visualization of the cell types' proportions**\
```{r, echo = TRUE, message = FALSE, warning = FALSE, fig.show="hold", out.width="50%"}
par(mar = c(5,4,4,8))

cs_reshaped1 <- melt(csDarmanis[1:52,], id = "columns")
cs_reshaped2 <- melt(csDarmanis[53:102,], id = "columns")

ggplot(cs_reshaped1, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
  scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
  theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('ALS')

ggplot(cs_reshaped2, aes(fill = Var2, x = Var1, y = value)) +
  geom_bar(position="stack", stat="identity") +
  scale_fill_discrete(name = "Cell types", labels = c("Oligodendrocytes", "Astrocytes", "OPC",
                                                      "Microglia", "Neurons", "Endothelial")) +
    theme(axis.text.x = element_text(size=10, angle = 90),
        axis.title = element_blank(),
        plot.title = element_text(size = 22, face = "bold"),
        legend.position = "bottom") + 
  ggtitle('Control')

```

***
***

**Statisical significance of the cell-types between the conditions**\
To conclude, the table below depicts the p-value of each Student t-test between the two conditions for each cell-type. As one might notice, the results from the NNLS and cibersort methods converge to the same conclusion regarding the statistical significance for the following cell-types: astrocytes, endothelial cells and neurons. On the other hands, the results from dtangle method suggest that only the proportions of neurons are not signficantly different between the conditions. This is not what would be expected since the ALS is known for altering motor neurons function [10]. One reason for that would be that the inputs of the method dtangle are not optimized. Nevertheless, more deconvolution methods exist and might be performed in order to have better insight of the results and make more appropriate comparisons. The above analysis thus suggests that the proportion of neurons between the two conditions is different. To strenghten this conclusion, one might, for instance, apply other gene specific markers datasets.\
```{r, echo = TRUE, message = FALSE, warning = FALSE}
dt_t <- t(dt)
prop_t <- t(prop)
cs_t <- t(csDarmanis)

test1 <- list()
for (i in 1:nrow(dt_t)) {
  ttest <- t.test(dt_t[i,1:52], dt_t[i,53:102])$p.value
  test1 <- append(test1, ttest)
}
names(test1) <- rownames(dt_t)

test2 <- list()
for (i in 1:nrow(prop_t)) {
  ttest <- t.test(prop_t[i,1:52], prop_t[i,53:102])$p.value
  test2 <- append(test2, ttest)
}
names(test2) <- rownames(prop_t)

test3 <- list()
for (i in 1:nrow(cs_t)) {
  ttest <- t.test(cs_t[i,1:52],cs_t[i,53:102])$p.value
  test3 <- append(test3, ttest)
}
names(test3) <- rownames(cs_t)

df <- as.data.frame(matrix(unlist(test1), nrow=length(test1), byrow = T))
df$V2 <- matrix(unlist(test2), nrow=length(test2), byrow = T)
df$V3 <- matrix(unlist(test3), nrow=length(test3), byrow = T)

df <- t(df)

colnames(df) <- c("Oligodendrocytes", "Astrocytes", "OPC", "Microglia", "Neurons", "Endothelial")
rownames(df) <- c("dtangle", "NNLS", "cibersort")

knitr::kable(df, format = "simple", align = "cccccc")
```

***
***

##References

[Reference for the approach](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008120)\
[Reference of the marker genes dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE67835)\
[1](https://www.ninds.nih.gov/Disorders/Patient-Caregiver-Education/Fact-Sheets/Amyotrophic-Lateral-Sclerosis-ALS-Fact-Sheet)\
[2](https://www.biorxiv.org/content/10.1101/2020.01.19.910976v1.full)\
[3](https://journals.plos.org/plosone/article?id=10.1371%2Fjournal.pone.0215987)\
[4](https://bioconductor.org/packages/release/bioc/html/edgeR.html)\
[5](https://www.pnas.org/content/112/23/7285)\
[6](https://academic.oup.com/bioinformatics/article/35/12/2093/5165376)\
[7](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0006098#s5)\
[8](https://www.biorxiv.org/content/10.1101/633958v2.full)\
[9](https://pubmed.ncbi.nlm.nih.gov/25822800/)\
[10](https://pubmed.ncbi.nlm.nih.gov/17015226/)\


##Acknowledgements

I thank Dr. Hubert Rehrauer for his assistance and the Functional Genomics Center Zurich for providing the dataset of expression values.





