---
title: "Cell type deconvolution"
author: "RA ([numpde](https://github.com/numpde/))"
date: "2020-Dec-26"
output: 
  html_document:
    toc: true
    toc_depth: 2
editor_options:
 chunk_output_type: console
---

## Introduction

(TODO)

## Pertinent links

- [Source code](https://github.com/numpde/sta426_deconvo/tree/main/code/sandbox_ra/20201226-AllDeco),
  short link: 
  [http://bit.ly/deco_ra](http://bit.ly/deco_ra)

- [FGCZ bulk dataset](https://github.com/numpde/sta426_deconvo/tree/main/data/20201128-FGCZ)

- [Darmanis et al (2015)](https://github.com/numpde/sta426_deconvo/tree/main/data/20201206-RefDatasets/2015-Darmanis) 
  reference scRNA dataset

- [Allen Brain M1](https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x) 
  reference scRNA dataset

- [Analysis](https://github.com/numpde/sta426_deconvo/tree/main/code/sandbox_ra/20201220-Darmanis_vs_Allen) 
  of **Darmanis et al** in terms of **Allen Brain M1**

- Bisque 
  [paper](https://www.nature.com/articles/s41467-020-15816-6),
  [vignette](https://cran.r-project.org/web/packages/BisqueRNA/vignettes/bisque.html),
  [manual](https://cran.r-project.org/web/packages/BisqueRNA/BisqueRNA.pdf),
  [github](https://github.com/cozygene/bisque),
  [ReferenceBasedDecomposition](https://rdrr.io/cran/BisqueRNA/src/R/reference_based.R)
  (uses 
  [limSolve](https://cran.r-project.org/web/packages/limSolve/limSolve.pdf))

- MuSiC
  [paper](https://www.nature.com/articles/s41467-018-08023-x),
  [tutorial](https://xuranw.github.io/MuSiC/articles/MuSiC.html),
  [github](https://github.com/xuranw/MuSiC),
  [music_prop](https://rdrr.io/github/xuranw/MuSiC/src/R/utils.R)




## Preliminaries {.tabset .tabset-fade .tabset-pills}

### Setup    

```{r setup, include=TRUE}
suppressPackageStartupMessages({
  requireNamespace("magrittr")
  library(dplyr)
  library(ggplot2)

  requireNamespace("pheatmap")
  requireNamespace("stringr")
  requireNamespace("reshape2")

  # avoid importing `exprs` that leads to clashes
  requireNamespace("rlang")

  # https://vroom.r-lib.org/articles/vroom.html
  # install.packages("vroom")
  requireNamespace("vroom")

  # install.packages("pathlibr")
  requireNamespace("pathlibr")

  # Provides `ExpressionSet` data structure
  # BiocManager::install("Biobase")
  requireNamespace("Biobase")

  # Deconvolution package
  # install.packages("BisqueRNA")
  requireNamespace("BisqueRNA")

  # # https://xuranw.github.io/MuSiC/articles/MuSiC.html
  # install.packages("devtools")
  # devtools::install_github("xuranw/MuSiC")
  requireNamespace("xbioc") # For MuSiC
  requireNamespace("MuSiC")

  # install.packages("kableExtra")
  requireNamespace("kableExtra")
})
```


```{r}
sessionInfo()
```



### Paths

```{r}
BASEPATH <- pathlibr::Path$new(".")
```

```{r}
out_file <- (function(.) BASEPATH$join("./b_output")$join(.)$show)
dir.create(out_file(""), showWarnings = FALSE)
```

```{r}
path_to <- (function(.) Sys.glob(BASEPATH$join("../../..")$join(.)$show))
```

### Random-seed

```{r}
set.seed(43)
```

### Row names, etc.

```{r}
# Use first column as index
by_col1 <- (function(.) tibble::column_to_rownames(., colnames(.)[1]))
# Use index as new column `name`
ind2col <- (function(., name) tibble::rownames_to_column(., var = name))
```

```{r}
group_by_and_sum <-
  function(., column_name) {
    (.) %>%
      dplyr::group_by(!!rlang::sym(column_name)) %>%
      dplyr::summarise_all(sum) %>%
      tibble::column_to_rownames(column_name)
  }
```

```{r}
fix_names <- function(.) {
  (.) %>%
    # https://stackoverflow.com/a/55184433
    rlang::set_names(stringr::str_replace(names(.), "cell type", "celltype"))
}
```

```{r}
col_norm2 <- (function(.) t(t(.) / sqrt(colSums((.)**2))))
```

### File I/O

The function `utils::read.delim` is too slow
to read wide tables (Allen Brain M1 has >70k cells).

```{r}
Sys.setenv("VROOM_CONNECTION_SIZE" = 131072 * 32)
from_csv <- function(.) {
  vroom::vroom(., del = "\t") %>%
    by_col1() %>%
    suppressMessages()
}
```

A call to `to_csv` should be preceded by
`ind2col("name for index")`
to write the row names.

```{r}
to_csv <- function(., f) {
  vroom::vroom_write(., out_file(f), delim = "\t")
}
```

### Plotting

```{r}
ggplot2::theme_set(theme_light(base_size = 15))
```

```{r}
kable <- function(.) {
  kableExtra::kbl(.) %>%
    kableExtra::kable_paper("hover", full_width = F)
}
```

```{r}
# Wide plots (inches?)
WIDTH1 <- 20
HEIGHT1 <- 3
```

```{r}
hist_theme <-
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.ticks.x = element_blank(),
  )
```

```{r}
save_barchart <-
  function(., filename) {
    as.data.frame(.) %>%
      ind2col("celltype") %>%
      reshape2::melt(
        id = "celltype",
        var = "sample",
        value.name = "y"
      ) %>%
      {
        ggplot(., aes(x = sample, y = y, fill = celltype)) +
          geom_bar(stat = "identity") +
          scale_fill_brewer(palette = "Paired") +
          ylim(0, 1) +
          hist_theme +
          theme(
            legend.title = element_blank(),

            axis.text.y = element_blank(),

            axis.title.x = element_blank(),
            axis.text.x = element_text(
              a = -90,
              vjust = 0.5,
              hjust = 0,
              size = 10,
            )
          ) +
          ggsave(
            filename = out_file(filename),
            width = WIDTH1,
            height = HEIGHT1,
            device = "png"
          )
      } %>%
      suppressWarnings()
    return(.)
  }
```

```{r}
save_heatmap <-
  function(., filename) {
    as.data.frame(.) %>%
      pheatmap::pheatmap(
        filename = out_file(filename),
        cluster_rows = FALSE,
        cluster_cols = FALSE,
        fontsize_row = 14,
        fontsize_col = 10,
        width = WIDTH1,
        height = HEIGHT1
      )
    return(.)
  }
```





## Data sources {.tabset .tabset-fade .tabset-pills}


### Bulk expression: FGCZ

```{r, cache=TRUE}
fgcz_data <-
  path_to("data/20201128-FGCZ/*count.zip") %>%
  unz(., unzip(., list = TRUE)$Name) %>%
  from_csv() %>%
  # Collapse ENSG IDs by gene_name:
  group_by_and_sum("gene_name")

fgcz_meta <- from_csv(path_to("data/20201128-FGCZ/*infos.tsv"))
```

This will be used later.

```{r}
assert_fgcz_meta_order <-
  function(.) {
    stopifnot(all(rownames(.) == rownames(fgcz_meta)))
    return(.)
  }
```

```{r}
t(fgcz_data) %>%
  assert_fgcz_meta_order() %>%
  invisible()
```

### Ref scRNA: Darmanis

```{r, cache=TRUE}
darm_data <- from_csv(path_to("data/*/2015-Darmanis/b*/data.csv.gz"))
darm_meta <- from_csv(path_to("data/*/2015-Darmanis/b*/meta.csv.gz")) %>%
  fix_names()
```

```{r}
exclude_celltypes <- c("fetal_quiescent", "fetal_replicating", "hybrid")
```

Plot cell type counts.

```{r}
darm_meta$celltype %>%
  data.frame(x = .) %>%
  ggplot(aes(x = x, fill = if_else(x %in% exclude_celltypes, "drop", "keep"))) +
  geom_bar() +
  ggtitle("Cell types in the 'Darmanis' reference dataset") +
  scale_y_log10() +
  hist_theme +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x = element_text(a = 45, hjust = 1)) +
  theme(legend.title = element_blank())
```

Remove unnecessary cell types
from the scRNA reference dataset.

```{r, cache=TRUE}
darm_meta <-
  darm_meta %>%
  dplyr::filter(!(celltype %in% exclude_celltypes))

darm_data <-
  darm_data %>%
  dplyr::select_if(names(.) %in% rownames(darm_meta))

stopifnot(285 == nrow(darm_meta))
stopifnot(285 == ncol(darm_data))
```


### Ref scRNA: Allen Brain M1

```{r, cache=TRUE}
abm1_meta <- from_csv(path_to("data/*/2019-AllenBrain-M1/b*/meta.csv*"))
abm1_data <- from_csv(path_to("data/*/2019-AllenBrain-M1/b*/data.csv*"))
```

Some sanity checks.

```{r}
# Number of single cells
stopifnot(ncol(abm1_data) > 70000)
stopifnot(nrow(abm1_meta) > 70000)
# Number of genes and some examples
stopifnot(nrow(abm1_data) == 141)
stopifnot(all(c("PIK3CD", "WNT4", "LDLRAP1") %in% rownames(abm1_data)))
```

Drop samples with too little expression,
otherwise Bisque doesn't go through.

```{r, cache=TRUE}
abm1_data <-
  abm1_data %>%
  # Keep only genes common with the FGCZ dataset
  {
    (.)[rownames(.) %in% rownames(fgcz_data), ]
  } %>%
  # Drop zero columns (dplyr variant too slow)
  {
    (.)[, colSums(.) != 0]
  }

abm1_meta <-
  abm1_meta %>%
  dplyr::filter(rownames(.) %in% colnames(abm1_data))
```

Plot cell type counts.

```{r}
data.frame(
  x = abm1_meta$celltype,
  donor = abm1_meta$donor
) %>%
  ggplot(aes(x = x, fill = donor)) +
  geom_bar(position = "dodge") +
  ggtitle("Cell types in the 'Allen Brain M1' reference dataset") +
  scale_fill_brewer(palette = "Set1") +
  scale_y_log10() +
  hist_theme +
  theme(axis.title.x = element_blank()) +
  theme(axis.title.y = element_blank())
```


### Marker genes

These marker genes 
[were extracted](https://github.com/numpde/sta426_deconvo/tree/main/data/20201206-RefDatasets/2019-AllenBrain-M1)
from the
[Allen Brain M1](https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x)
scRNA dataset.
Technically, these should be the same as 
in `abm1_data` before filtering.
Just in case, we subset them the maximal common subset.


```{r, cache=TRUE}
abm1_markergenes <-
  path_to("data/*/*AllenBrain-M1/b*/marker_genes.csv") %>%
  from_csv() %>%
  rownames() %>%
  {
    (.)[(.) %in% rownames(fgcz_data)]
  } %>%
  {
    (.)[(.) %in% rownames(abm1_data)]
  }

stopifnot(117 == length(abm1_markergenes))
```



## Deconvolution


### From Darmanis {.tabset .tabset-fade .tabset-pills}

In this section we use
**Darmanis et al**
as the reference dataset.

#### Bisque {#fgcz_darm_bisque}

Repackage the data
following the 
[Bisque vignette](https://cran.r-project.org/web/packages/BisqueRNA/vignettes/bisque.html).

```{r}
fgcz_eset <- Biobase::ExpressionSet(
  # Subset to marker genes for residual norm computation
  assayData = as.matrix(fgcz_data[abm1_markergenes, ])
)
```

```{r}
darm_eset <- Biobase::ExpressionSet(
  # Expression data
  assayData = as.matrix(darm_data),
  # Metadata
  phenoData = Biobase::AnnotatedDataFrame(
    data = data.frame(
      row.names = rownames(darm_meta),
      cellType = darm_meta$celltype,
      SubjectName = darm_meta$experiment_sample_name,
      check.names = FALSE,
      check.rows = FALSE,
      stringsAsFactors = FALSE
    ),
    varMetadata = data.frame(
      row.names = c("cellType", "SubjectName"),
      labelDescription = c("cellType", "SubjectName")
    )
  )
)
```

Deconvolution.

```{r, cache=TRUE}
bisque_report <-
  BisqueRNA::ReferenceBasedDecomposition(
    # BULK DATA
    bulk.eset = fgcz_eset,
    # REFERENCE
    sc.eset = darm_eset,
    #
    use.overlap = FALSE,
    markers = abm1_markergenes,
    verbose = FALSE
  )
```

```{r}
# Fields of the Bisque result
bisque_report %>%
  summary() %>%
  kable()
```


Bisque returns proportions 
that sum to one:

```{r}
stopifnot(max(abs(1 - colSums(bisque_report$bulk.props))) <= 1e-10)
```

We undo that using the residual.
The meaning of `rnorm` is not clear
from the
[manual](https://cran.r-project.org/web/packages/BisqueRNA/BisqueRNA.pdf),
so this might be incorrect:

```{r}
explained_fractions <-
  bisque_report %>% {
    1 - (sqrt((.)$rnorm) / sqrt(colSums(fgcz_data[(.)$genes.used, ]**2)))
  }
```

Rescale proportions to 
the "explained fractions".

```{r}
fgcz_darm_bisque <-
  t(t(bisque_report$bulk.props) * explained_fractions)
```

Cluster bulk samples by composition.

```{r}
samples_order <-
  fgcz_darm_bisque %>%
  {
    (.)[, hclust(dist(t(.)))$order]
  } %>%
  colnames()
```

Save to disk
and
visualize inferred cell type composition
by bulk sample.

```{r}
visualize_and_save <-
  function(., prefix = deparse(substitute(.))) {
    as <- {
      function(ext) paste(prefix, ext, sep = "_")
    }

    (.)[, samples_order] %>%
      #
      save_heatmap(as("heat.png")) %>%
      save_barchart(as("bars.png")) %>%
      #
      as.data.frame() %>%
      ind2col("celltype") %>%
      to_csv(as("data.csv"))
  }
```

```{r}
visualize_and_save(fgcz_darm_bisque)
```

![](`r out_file("fgcz_darm_bisque_heat.png")`)
![](`r out_file("fgcz_darm_bisque_bars.png")`)


The cell type proportions [returned by Bisque](#fgcz_darm_bisque)
vary rather more than expected.
The samples come from four
different hospitals 
(the field `fgcz_meta$Source`),
which could introduce the strongest batch effect.
We can also compare that 
to the [RIN](https://en.wikipedia.org/wiki/RNA_integrity_number)
(RNA integrity number).
Age could be another important factor.

```{r}
PC1_versus <-
  function(., on_the_y_axis) {
    prcomp(t(.))$x %>%
      as.data.frame() %>%
      assert_fgcz_meta_order() %>%
      cbind(q = (2 * rank((.)$PC1) / length((.)$PC1) - 1)) %>%
      cbind(c = fgcz_meta$Condition) %>%
      cbind(s = fgcz_meta$Source) %>%
      cbind(Age = fgcz_meta$Age) %>%
      cbind(RIN = fgcz_meta$RIN) %>%
      cbind(y = (.)[[on_the_y_axis]]) %>%
      {
        ggplot(., aes(x = q, y = y, shape = c, color = s)) +
          labs(
            x = "Deviation from the average composition (~PC1)",
            y = on_the_y_axis,
            color = "Sample source",
            shape = "Condition"
          ) +
          geom_point(size = 5, alpha = 0.2 + abs((.)$q)) +
          scale_color_brewer(palette = "Set1")
      }
  }
```

Principal components of cell type proportions.

```{r}
fgcz_darm_bisque %>%
  PC1_versus("PC2")
```

```{r}
fgcz_darm_bisque %>%
  PC1_versus("RIN")
```

```{r}
fgcz_darm_bisque %>%
  PC1_versus("Age")
```



#### MuSiC

```{r, cache=TRUE}
# Prevent errors from MuSiC
exprs <- xbioc::exprs
pVar <- xbioc::pVar

music_report <-
  MuSiC::music_prop(
    bulk.eset = fgcz_eset,
    sc.eset = darm_eset,
    clusters = darm_meta$celltype,
    samples = names(darm_data)
  ) %>%
  suppressMessages()
```

```{r}
music_report %>%
  summary() %>%
  kable()
```

Estimate cell type proportions (celltype x sample).

```{r}
fgcz_darm_music <- t(music_report$Est.prop.weighted)
```

```{r}
visualize_and_save(fgcz_darm_music)
```

![](`r out_file("fgcz_darm_music_heat.png")`)
![](`r out_file("fgcz_darm_music_bars.png")`)

Principal components of cell type proportions.

```{r}
fgcz_darm_music %>%
  PC1_versus("PC2")
```

```{r}
fgcz_darm_music %>%
  PC1_versus("RIN")
```

```{r}
fgcz_darm_music %>%
  PC1_versus("Age")
```

We would like to know
whether there is 
the cell type composition differs 
in ALS vs Control samples.
First,
the following shows the distribution
of the first principal component
(of cell type proportions)
separated by sample source.

```{r, eval=TRUE}
fgcz_darm_music %>%
  {
    as.data.frame(prcomp(t(.))$x)
  } %>%
  assert_fgcz_meta_order() %>%
  cbind(c = fgcz_meta$Condition) %>%
  cbind(s = fgcz_meta$Source) %>%
  cbind(a = fgcz_meta$Age) %>%
  {
    ggplot(., aes(x = PC1, fill = s, linetype = s)) +
      labs(fill = "Sample source") +
      guides(linetype = FALSE) +
      geom_density(
        position = "identity", alpha = 0.4
      ) +
      geom_histogram(
        aes(y = stat(density) / 5),
        position = "dodge", bins = 20, color = 1, linetype = 1
      ) +
      hist_theme +
      theme(axis.text.y = element_blank()) +
      theme(
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
      ) +
      scale_fill_brewer(palette = "Set1")
  }
```

**(IN PROGRESS)**




```{r, eval=FALSE}
library(SingleCellExperiment)
library(scater)
factor_name <- "Source"
fgcz_darm_music %>%
  {
    list(logcounts = (.))
  } %>%
  SingleCellExperiment(assays = .) %>%
  scater::runPCA(ncomponents = 2) %>%
  SingleCellExperiment::reducedDim() %>%
  as.data.frame() %>%
  setNames(c("x", "y")) %>%
  cbind(s = fgcz_meta$Source) %>%
  cbind(c = fgcz_meta$Condition) %>%
  cbind(a = fgcz_meta$Age) %>%
  {
    ggplot(., aes(x = x, y = y, color = c)) +
      geom_point() +
      labs(
        color = "Sample source",
        shape = "Condition"
      ) +
      geom_point(size = 5)
    # scale_color_brewer(palette = "Set1")
  }
```

```{r, eval=FALSE}
# MuSiC with ABM1
music_report <-
  MuSiC::music_prop(
    bulk.eset = fgcz_eset,
    sc.eset = abm1_eset,
    clusters = abm1_meta$celltype,
    samples = names(abm1_data)
  ) %>%
  suppressMessages()

print(as.data.frame(names(music_report)))
```



#### Baseline NNLS

For the code see 
[here](https://github.com/numpde/sta426_deconvo/blob/main/code/sandbox_ra/20201206-NNLS_Darmanis/a_deconvolution0.py).

```{r}
fgcz_darm_nnls <-
  path_to("code/*/*-NNLS_Darmanis/a_*/celltypes.csv") %>%
  # The first column is not unique
  # Keep header as is, e.g. "H-0C083"
  utils::read.delim(., check.names = F) %>%
  # `cell type` ~> `celltype`
  fix_names() %>%
  # Collapse cell types:
  group_by_and_sum("celltype")
```

```{r}
visualize_and_save(fgcz_darm_nnls)
```

![](`r out_file("fgcz_darm_nnls_heat.png")`)
![](`r out_file("fgcz_darm_nnls_bars.png")`)


```{r}
fgcz_darm_nnls %>%
  PC1_versus("PC2")
```

```{r}
fgcz_darm_nnls %>%
  PC1_versus("RIN")
```

```{r}
fgcz_darm_nnls %>%
  PC1_versus("Age")
```


These results are similar enough to MuSiC's
to warrant a closer look.

```{r}
# Check that samples are arranged consistently
stopifnot(all(colnames(fgcz_darm_music) == colnames(fgcz_darm_nnls)))

colSums(col_norm2(fgcz_darm_music) * col_norm2(fgcz_darm_nnls)) %>%
  data.frame(x = .) %>%
  ggplot(aes(x = x)) +
  geom_histogram() +
  hist_theme +
  labs(x = "Cosine similarity (NNLS vs MuSiC)")
```



### From Allen Brain M1

In this section we use
**Allen Brain M1**
as the reference dataset.

Repackage:

```{r}
abm1_eset <- Biobase::ExpressionSet(
  # Expression data
  assayData = as.matrix(abm1_data),
  # Metadata
  phenoData = Biobase::AnnotatedDataFrame(
    data = data.frame(
      row.names = rownames(abm1_meta),
      cellType = abm1_meta$celltype,
      SubjectName = abm1_meta$donor,
      check.names = FALSE,
      check.rows = FALSE
    ),
    varMetadata = data.frame(
      row.names = c("cellType", "SubjectName"),
      labelDescription = c("cellType", "SubjectName")
    )
  )
)
```

Deconvolution:

```{r, cache=TRUE}
bisque_report <-
  BisqueRNA::ReferenceBasedDecomposition(
    # BULK DATA
    bulk.eset = fgcz_eset,
    # REFERENCE
    sc.eset = abm1_eset,
    #
    use.overlap = FALSE,
    markers = abm1_markergenes,
    verbose = FALSE
  )
```

```{r}
bisque_report$bulk.props %>%
  visualize_and_save("fgcz_abm1_bisque")
```

![](`r out_file("fgcz_abm1_bisque_heat.png")`)
![](`r out_file("fgcz_abm1_bisque_bars.png")`)





## Acknowledgements

(TODO)


***


```{r include=FALSE}
knitr::knit_exit()
```

## DWLS

```{r}
# https://github.com/dtsoucas/DWLS/blob/master/Manual.docx
BiocManager::install("MAST")
remotes::install_bitbucket("yuanlab/dwls")
requireNamespace("DWLS")
```

```{r}
x <- MAST::FromMatrix(
  exprsArray = as.matrix(darm_data),
  cData = darm_meta,
  class = "SingleCellAssay",
  check_sanity = FALSE
)

x <-
  DWLS::buildSignatureMatrixMAST(
    scdata = as.matrix(darm_data),
    id = setNames(as.list(darm_meta[["celltype"]]), rownames(darm_meta))
  )

require("ROCR")
require("MAST")
scdata_dummy <- matrix(sample.int(100, size = 4 * 15, replace = T), nrow = 5, ncol = 4)
rownames(scdata_dummy) <- paste("Gene", 1:nrow(scdata_dummy), sep = "")
colnames(scdata_dummy) <- paste("Cell", 1:ncol(scdata_dummy), sep = "")
y <- DWLS::DEAnalysisMAST(scdata = scdata_dummy, id = c("A", "B", "A", "B"), path = "./here.dat")
# Fails with:
#  Error in .subset2(x, i, exact = exact) : invalid subscript type 'S4'
```


[//]: # XX -- TODO

```{r, eval=FALSE, echo=FALSE}
library(SingleCellExperiment)
library(scater)
factor_name <- "Source"
fgcz_data %>%
  edgeR::cpm(.) %>%
  {
    (.)[abm1_markergenes, ]
  } %>%
  {
    list(logcounts = log(1 + .))
  } %>%
  SingleCellExperiment(assays = .) %>%
  scater::runMDS() %>%
  SingleCellExperiment::reducedDim() %>%
  as.data.frame() %>%
  setNames(c("x", "y")) %>%
  ggplot(aes(x = x, y = y, color = fgcz_meta[[factor_name]])) +
  labs(color = factor_name) +
  geom_point(size = 5)
```

```{r, eval=FALSE, echo=FALSE}
fgcz_abm1 %>%
  t() %>%
  as.data.frame() %>%
  dplyr::mutate(condition = fgcz_meta$Condition) %>%
  {
    MASS::lda(formula = condition ~ ., data = (.))
  }
```
